feat:  Add system watchdog configuration doc (#48465)
kubernetes/website: content/en/docs/reference/node/systemd-watchdog.md

---
content_type: "reference"
title: Kubelet Systemd Watchdog
weight: 80
---

{{< feature-state feature_gate_name="SystemdWatchdog" >}}

> Using the watchdog can improve the reliability of the kubelet.

Reviewer: ```suggestion
On Linux nodes, Kubernetes {{< skew currentVersion >}} supports integrating with
[systemd](https://systemd.io/) to allow the operating system supervisor to recover
a failed kubelet. This integration is not enabled by default.

```

It can be used as an alternative to periodically requesting
the kubelet's `/healthz` endpoint for health checks. If the kubelet
does not respond to the watchdog within the timeout period, the watchdog
will kill the kubelet.

The systemd watchdog works by requiring the service to periodically send

> a "keep-alive" signal to the systemd daemon. If the signal is not received

Reviewer: ```suggestion
a _keep-alive_ signal to the systemd process. If the signal is not received
```
Although intuition might make you think otherwise, systemd is not really a daemon.

within a specified timeout period, the service is considered unresponsive
and is terminated. The service can then be restarted according to the configuration.

> ### How to use

Reviewer: ```suggestion
## Configuration
```

> and is terminated. The service can then be restarted according to the configuration.

Author: ok

> ### Who to use

Reviewer 2: Was this section title intended to be 'How to use' instead?

```suggestion
### How to use
```

> Using the systemd watchdog only requires configuring the WatchdogSec= parameter

Reviewer 2: ```suggestion
Using the systemd watchdog only requires configuring the `WatchdogSec` parameter
```

> Using the systemd watchdog only requires configuring the `WatchdogSec` parameter

Author: got

> ```sh

Reviewer: This is not a shell script; omit the `sh` here.

[Service]
WatchdogSec=30s
```

> `WatchdogSec=30s` indicates a timeout of 30 seconds. In the kubelet program,

Reviewer: ```suggestion
Setting `WatchdogSec=30s` indicates a service watchdog timeout of 30 seconds. Within the kubelet
```

> the `sd_notify` function is called at intervals of `WatchdogSec/2` to send

Reviewer: (nit)
```suggestion
the kubelet invokes the `sd_notify()` function, at intervals of `WatchdogSec` ÷ 2, to send
```

> "WATCHDOG=1" for the keep-alive operation. If the watchdog is not fed

Reviewer: (nit)
```suggestion
`WATCHDOG=1` (a keep-alive message). If the watchdog is not fed
```

"WATCHDOG=1" for the keep-alive operation. If the watchdog is not fed

> within the timeout period, the kubelet will be killed. Setting `Restart=`

Reviewer 2: ```suggestion
within the timeout period, the kubelet will be killed. Setting `Restart`
```

is automatically restarted.

> Explanation about the `WatchdogSec` configuration:

Reviewer: ```suggestion
Some details about the systemd configuration:
```

> 1. Not configuring it or setting it to 0 means the watchdog is not enabled.

Reviewer: ```suggestion
1. If you set the systemd value for `WatchdogSec` to 0, or omit setting it, the systemd watchdog is not
   enabled for this unit.
```

> 1. Not configuring it or setting it to 0 means the watchdog is not enabled.

Reviewer 3: The minimum value is `1s`, but it can be set to `0` ...
This is a little bit confusing.
So please be careful with the text.

Author: 0 indicates disabling the watchdog. Although `WatchdogSec` can be set to less than 1 second, Kubelet enforces a minimum value of 1 second. This can be confusing, so I will clarify it further.

> 2. The minimum value for WatchdogSec is currently 1 second to avoid
>    the kubelet being killed unexpectedly.

Reviewer 3: ```suggestion
2. The minimum value for WatchdogSec is currently 1 second to prevent
   the kubelet from being killed unexpectedly.
```

3. The maximum value should not exceed 10 minutes,

>    because the syncLoop health check considers the kubelet unhealthy

Reviewer 3: If you want to highlight `syncLoop`, you'd better explain it.
If it is irrelevant for users to understand the mechanism, we should avoid referring to the function names.
Your audience are k8s users, not developers.

Author: Thank you, you are absolutely right. I have decided to remove the function names.

> 2. The default minimum value for `WatchdogSec` in kubelet is currently 1 second 
>    to prevent the kubelet from being killed unexpectedly. Even though `WatchdogSec`
>    in systemd configuration can be set to less than 1 second, 
>    Kubelet enforces a minimum of 1 second.

Reviewer: Is this right?
```suggestion
2. The kubelet supports a minimum watchdog period of 1.0 seconds; this is to prevent the kubelet
   from being killed unexpectedly. You can set the value of `WatchdogSec` in a systemd unit definition
   to a period shorter than 1 second, but Kubernetes does not support any shorter interval.  
   The timeout does not have to be a whole integer number of seconds.
```

Reviewer: BTW the trailing whitespace on (new) line 41 is intentional.

> 3. A good practice is to set `WatchdogSec` to no more than 10 minutes, 
>    although this is not mandatory.

Reviewer: ```suggestion
3. The Kubernetes project suggests setting `WatchdogSet` to approximately a 5 second period.
   Periods longer than 10 minutes are supported but explicitly **not** recommended.
```

> ### Example Configuration

Reviewer: ```suggestion
## Example unit configuration
```

> ```Ini

Reviewer: Don't use `Ini` here; systemd uses a similar but different format AIUI.

[Unit]
Description=kubelet: The Kubernetes Node Agent
Documentation=https://kubernetes.io/docs/home/
Wants=network-online.target
After=network-online.target

[Service]
ExecStart=/usr/bin/kubelet

> WatchdogSec=30s

Reviewer: Ideally, highlight this line - see https://jacobtomlinson.dev/posts/2024/how-to-highlight-lines-in-a-hugo-code-block/ for more detail on how to do that.

Author: I tried some syntax from the documentation, but none of it worked. For now, I'll use a comment to emphasize it.

Reviewer: No worries, we can fix that post merge.

Restart=on-failure
StartLimitInterval=0
RestartSec=10

[Install]
WantedBy=multi-user.target
```

> For more detailed configuration, please refer to the

Reviewer 3: ```suggestion
For more details about systemd configuration, please refer to the
```

> For more details about systemd configuration, please refer to the

Reviewer: ```suggestion

## {{% heading "whatsnext" %}}

For more details about systemd configuration, refer to the
```
